dali/kernels/slice/slice_gpu.cuh

```diff
@@ -137,65 +132,39 @@ __device__ void SliceFuncNoPad(OutputType *__restrict__ out, const InputType *__
  * @brief General algorithm that allows for padding in any dimension
  * @remarks `in` refers to the beginning of the input (not the slice anchor)
- * @remarks `AllDims=true` means that Dims refer to the actual number of dimensions,
- *           meaning we haven't skipped last dimensions that have same input and output strides
  */
-template <int Dims, typename OutputType, typename InputType, bool AllDims = true>
+template <int Dims, typename OutputType, typename InputType>
 __device__ void SliceFunc(OutputType *__restrict__ out, const InputType *__restrict__ in,
                           const fast_div<uint64_t> *out_strides, const int64_t *in_strides,
                           const int64_t *out_shape, const int64_t *in_shape, const int64_t *anchor,
-                          const OutputType *__restrict__ fill_values, int channel_dim,
-                          uint64_t offset, uint64_t block_end) {
-  if (Dims > 1 && anchor[Dims - 1] == 0 && in_shape[Dims - 1] == out_shape[Dims - 1] &&
-      channel_dim != Dims - 1) {
-    const int NextDims = Dims > 1 ? Dims - 1 : 1;
-    SliceFunc<NextDims, OutputType, InputType, false>(out, in, out_strides, in_strides, out_shape,
-                                                      in_shape, anchor, fill_values, channel_dim,
-                                                      offset, block_end);
-    return;
-  }
-
-  constexpr int LastDim = Dims - 1;
-  int64_t inner_in_anchor = anchor[LastDim];
-  int64_t inner_in_extent = in_shape[LastDim];
-  if (!AllDims) {  // if we fused dimensions, adjust inner dimension's anchor and extent
-    inner_in_anchor = anchor[LastDim] * in_strides[LastDim];
-    inner_in_extent = Dims > 1 ? in_strides[LastDim - 1] : in_shape[LastDim] * in_strides[LastDim];
-  }
-
+                          const int64_t *step, const OutputType *__restrict__ fill_values,
+                          int channel_dim, uint64_t offset, uint64_t block_end) {
   for (; offset < block_end; offset += blockDim.x * PackedBuffer<OutputType>::kCapacity) {
     PackedBuffer<OutputType> result;
 
-    uint64_t i;
-    #ifndef __clang__
-    #pragma unroll
-    #endif
-    for (i = 0; i < PackedBuffer<OutputType>::kCapacity; i++) {
+    uint64_t i = 0;
+#ifndef __clang__
+#pragma unroll
+#endif
+    for (; i < PackedBuffer<OutputType>::kCapacity; i++) {
       uint64_t idx = offset + i;
-      if (idx >= block_end) break;
+      if (idx >= block_end)
+        break;
 
-      // If no dimensions were skipped (AllDims=true) we can avoid division in the last dimension,
-      // because know the strides are 1 (or we treat them as 1 if we fused dimensions)
       int i_c = 0;
       int i_d;
       bool out_of_bounds = false;
       uint64_t in_idx = 0;
 
-      #pragma unroll
-      for (int d = 0; d < Dims - 1; d++) {
-        i_d = div_mod(idx, idx, out_strides[d]);
+#pragma unroll
+      for (int d = 0; d < Dims; d++) {
+        i_d = anchor[d] + div_mod(idx, idx, out_strides[d]) * step[d];
```

Comment:
I don't think this simplification is necessary - we can keep the last iteration outside and avoid one div_mod, which isn't cheap.